synapse/handlers/e2e_keys.py

```diff
@@ -511,6 +513,262 @@ def upload_signing_keys_for_user(self, user_id, keys):
 
         defer.returnValue({})
 
+    @defer.inlineCallbacks
+    def upload_signatures_for_device_keys(self, user_id, signatures):
+        """Upload device signatures for cross-signing
+
+        Args:
+            user_id (string): the user uploading the signatures
+            signatures (dict[string, dict[string, dict]]): map of users to
+                devices to signed keys
+        """
+        failures = {}
+
+        # signatures to be stored.  Each item will be a tuple of
+        # (signing_key_id, target_user_id, target_device_id, signature)
+        signature_list = []
+        # what devices have been updated, for notifying
+        self_device_ids = []
+
+        # split between checking signatures for own user and signatures for
+        # other users, since we verify them with different keys
+        if user_id in signatures:
+            self_signatures = signatures[user_id]
+            del signatures[user_id]
+            self_device_ids = list(self_signatures.keys())
+            try:
+                # get our self-signing key to verify the signatures
```

Comment:
I haven't read the code in detail, but as a starting point for breaking up this function, can you not factor out the inside of this `try: except` block?

```python
try:
    signatures, failures = yield self._process_self_signatures(self_signatures))
    signature_list.extend(signatures)
    if failures:
        failures[user_id] = failures
except SynapseError as e:
    # etc
```

I'd also make `_process_self_signatures` start with:

```python
if not self_signatures:
    return [], None
```

so that you can remove the `if user_id in signatures` condition here.

(You can probably do something similar for the other_users case).